KS-18-03-tamagochi

DenisGubernatorov opened the pull request:
1) Общий класс TamagochiAnimal с пустой реализацией метода eat 2) Классы Fox, Dog, Cat наследники общего класса и переопределяющие метод eat

DenisGubernatorov committed:
KS-18-03-tamagochi
1) Общий класс TamagochiAnimal с пустой реализацией метода eat
2) Классы Fox, Dog, Cat наследники общего класса и переопределяющие метод eat

ievetrovdev commented on src/main/kotlin/lesson18/lesson18_task3.kt:
```diff
@@ -1 +1,36 @@
 package lesson18
+
+open class TamagochiAnimal(
```
Корректнее сделать класс абстрактным. Это позволит подчеркнуть, что данный класс не предназначен для создания экземпляров, а служит только для базового описания

ievetrovdev commented on src/main/kotlin/lesson18/lesson18_task3.kt:
```diff
@@ -1 +1,36 @@
+}
+
+class Fox(
+    override val name: String,
```
Переопределение этого свойства в данном случае избыточно и снижает читаемость кода.
Оно уже определено в родительском классе и будет доступно всем наследникам. Как правило переопределять нужно, когда мы хотим изменить модификатор доступа

ievetrovdev commented on src/main/kotlin/lesson18/lesson18_task3.kt:
```diff
@@ -1 +1,36 @@
+open class TamagochiAnimal(
+    open val name: String,
+) {
+    open fun eat() {}
```
В условии еще есть информация о том, что все звери могут спать. Это признак того, что для этого действия можно сделать реализацию по умолчанию

DenisGubernatorov committed:
KS-18-03-tamagochi

1) Класс TamagochiAnimal сделан абстрактным
2) Добавлен публичный метод sleep, с дефолтной реализацией

ievetrovdev commented on src/main/kotlin/lesson18/lesson18_task3.kt:
```diff
@@ -1 +1,43 @@
+abstract class TamagochiAnimal(
+    open val name: String,
+) {
+    open fun eat() {}
```
Еще момент: 
Поскольку еда у всех своя, лучше тут сделать абстрактную функцию и реализовывать ее в дочерних классах.

DenisGubernatorov committed:
1KS-18-03-tamagochi
1) Функция eat  к классе-родителе сделана абстрактной

ievetrovdev reviewed (approved):
Теперь все хорошо!
